### PR TITLE
Add a device parameter to RemoteModule

### PR DESCRIPTION
**Pull request by wayi1**
Summary:
Add a device parameter to RemoteModule, so it can be placed on any device
and not just CPU.

Original PR issue: RemoteModule enhancements #40550

Test Plan: buck test test/distributed/rpc:process_group_agent -- RemoteModule

Differential Revision: D23483803

### CODE REVIEW

File: torch/distributed/nn/api/remote_module.py [hunk 1]
```diff
@@ -96,6 +99,7 @@ def __init__(
                 >>> module_cls = MyModule
             args (Sequence, optional): args to be passed to ``module_cls``.
             kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+            device (torch.device): Device where the remote node runs the user-provided module.
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:102]: Reword this to: `Device on the destination worker where we'd like to place this module`

File: torch/distributed/nn/api/remote_module.py [hunk 2]
```diff
@@ -322,6 +326,7 @@ class RemoteModule(_RemoteModule):
             >>> module_cls = MyModule
         args (Sequence, optional): args to be passed to ``module_cls``.
         kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+        device (torch.device): Device where the remote node runs the user-provided module.
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:329]: same as above.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py [hunk 3]
```diff
@@ -208,6 +211,25 @@ def test_remote_parameters(self):
             self.assertEqual(len(param_rrefs), 1)
             self.assertTrue(torch.equal(param_rrefs[0].to_here(), _PARAM_VAL))
 
+    @dist_utils.dist_init
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/nn/api/remote_module_test.py:214]: You can just use the decorator `@skip_if_lt_x_gpu(1)` to run the test only on GPUs.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py [hunk 4]
```diff
@@ -208,6 +211,25 @@ def test_remote_parameters(self):
...
+
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        for remote_module in self._create_remote_module_iter(
+            dst_worker_name, modes=[ModuleCreationMode.MODULE_CTOR], device=device
+        ):
+            for param_rref in remote_module.remote_parameters():
+                param = param_rref.to_here()
+                if torch.cuda.is_available():
+                    self.assertEqual(param.device.type, "cuda")
+                    self.assertEqual(param.device.index, 0)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/nn/api/remote_module_test.py:228]: nit: Probably better to perform an RPC on the remote node and retrieve the device type from there. This just adds an extra safety net to ensure `to_here` isn't messing up something.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py [hunk 5]
```diff
@@ -208,6 +211,25 @@ def test_remote_parameters(self):
             self.assertEqual(len(param_rrefs), 1)
             self.assertTrue(torch.equal(param_rrefs[0].to_here(), _PARAM_VAL))
 
+    @dist_utils.dist_init
+    def test_device(self):
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/nn/api/remote_module_test.py:215]: Can we add a test for invalid devices (ex: "foo", "cuda:100", "cpu2" etc.) and validate that we raise appropriate exceptions.

File: torch/distributed/nn/api/remote_module.py [hunk 6]
```diff
@@ -164,7 +168,7 @@ def __init__(
 
         # Create the module on the remote side.
         self.module_rref = rpc.rpc_sync(
-            on, _create_module, (module_cls, args, kwargs, _module_interface_cls)
+            on, _create_module, (module_cls, args, kwargs, device, _module_interface_cls)
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:171]: It might be better to wrap device as `device = torch.device(device)` on the client side so that if we have an invalid device type an appropriate error is raised to the user. In the current scenario, we would invoke an RPC on the remote end and the `.to()` on the remote end would fail. The RPC framework would return that exception, but I'm not sure if it would be as clear as raising on the client side.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/distributed/nn/api/remote_module.py (hunk 6 above)
Comment (pritamdamania87) [reply torch/distributed/nn/api/remote_module.py:171]: Actually nvm, we should probably not do this. The reason being that sometimes a device might exist only on the remote node and not on the client and this would give us unintended errors.

Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:171]: Thanks for the comments. We can always revisit the API once our use cases become more clear. I believe RPC framework now is still in experimental phase, so the API here doesn't have to be finalized.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py (hunk 4 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/nn/api/remote_module_test.py:228]: Do you want me to add an RPC just in this test file or create a new method remote_device() in RemoteModule?

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py (hunk 5 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/nn/api/remote_module_test.py:215]: Added another test_invalid_device testcase.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py (hunk 4 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/nn/api/remote_module_test.py]: Just add an RPC in this test file to retrieve the remote device type.

File: torch/distributed/nn/api/remote_module.py [hunk 7]
```diff
@@ -35,6 +46,15 @@ def _create_module(module_cls, args, kwargs, module_interface_cls=None):
         )
     if module_interface_cls is not None:
         module = torch.jit.script(module)
+    if device:
+        d = torch.device(device)
+        if d.type == "cuda" and d.index >= torch.cuda.device_count():
+            raise RuntimeError(
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:52]: `torch.device(` would already cover any errors and throw an exception, why do we need this additional line? Also, is it sufficient to just call `.to()` here and would it throw the appropriate exception?  Or does `.to()` not throw an appropriate exception.

File: torch/distributed/nn/api/remote_module.py [hunk 8]
```diff
@@ -35,6 +46,15 @@ def _create_module(module_cls, args, kwargs, module_interface_cls=None):
...
+                "CUDA device index must be less than the number of available CUDA devices {}".format(
+                    torch.cuda.device_count()
+                )
+            )
+        module.to(d)
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:57]: Is this a false positive? Looks like it's complaining that this is not covered by tests? I'm assuming its probably because its happening on a different process?

File: torch/distributed/nn/api/remote_module.py [hunk 9]
```diff
@@ -322,6 +346,7 @@ class RemoteModule(_RemoteModule):
             >>> module_cls = MyModule
         args (Sequence, optional): args to be passed to ``module_cls``.
         kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+        device (str): Device on the destination worker where we‘d like to place this module.
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:349]: Consistent with rest of PyTorch (ex: .to()) docs we should make this `device (torch.device):` ....

File: torch/distributed/nn/api/remote_module.py (hunk 7 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:52]: For the input string like "cuda:100", `.to()` will not throw an exception until some real computation is deployed. Shall we move such check to module.to()?

File: torch/distributed/nn/api/remote_module.py (hunk 6 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py]: > Actually nvm, we should probably not do this. The reason being that sometimes a device might exist only on the remote node and not on the client and this would give us unintended errors.

Reverted this change after the second review.

File: torch/distributed/nn/api/remote_module.py (hunk 9 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:349]: Reverted the previous change from torch.device to str.

File: torch/distributed/nn/api/remote_module.py (hunk 7 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:52]: Please ignore the above comment. You are totally correct.

This is handled by correctly by PyTorch lib, confirmed by devgpu. Previously I didn't have the access to devgpu, so I used a Colab to test the error message, which may not have the most recent version installed.

File: torch/distributed/nn/api/remote_module.py (hunk 9 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:349]: Changed it back to torch.device.

File: torch/distributed/nn/api/remote_module.py (hunk 8 above)
Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py:57]: Yes, this actually is covered by the tests.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/nn/api/remote_module.py [hunk 10]
```diff
@@ -96,6 +110,7 @@ def __init__(
                 >>> module_cls = MyModule
             args (Sequence, optional): args to be passed to ``module_cls``.
             kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+            device (torch.device): Device on the destination worker where we‘d like to place this module.
```
Comment (pritamdamania87) [inline torch/distributed/nn/api/remote_module.py:113]: Forgot to mention this earlier, but it might be good to make `device` the second parameter (right after we specify the destination). So invocations would be like `RemoteModule("trainer0", "cuda:3", ...)`. We can probably make this API change since the feature hasn't been announced yet.

Comment (SciPioneer) [reply torch/distributed/nn/api/remote_module.py]: The only drawback here is the backward compatibility, because the previous second parameter is module_cls, which does not have a default value. Now this means device cannot have a default value.

This should be fine, since the constructor has only been used in a few unit tests.
